include/net/buf.h

```diff
@@ -1299,6 +1299,90 @@ struct net_buf *net_buf_frag_del_debug(struct net_buf *parent,
 struct net_buf *net_buf_frag_del(struct net_buf *parent, struct net_buf *frag);
 #endif
 
+/**
+ * @brief Copy len bytes from src starting from offset to dst buffer
+ *
+ * This routine assumes that dst is large enough to store @a len bytes
+ * starting from offset at src.
+ *
+ * @param dst Destination buffer
+ * @param dst_len Destination buffer max length
+ * @param src Source buffer that may be fragmented
+ * @param offset Starting point to copy from
+ * @param len Number of bytes to copy
+ * @return number of bytes copied if everything is ok
+ * @return -ENOMEM on error
+ */
+int net_buf_linearize(u8_t *dst, size_t dst_len,
```

Comment:
Might be friendlier to use `void *dst` to not require the caller do explicit typecasts if the destination is of some other type (say, a packed protocol struct).

Reply:
I'm ok with this in concept, however, the current API for net_pkt isn't designed this way.  See the existing APIs (which is what I based the additions on in net_buf):
linearize: https://github.com/zephyrproject-rtos/zephyr/blob/master/include/net/net_pkt.h#L1113
append: https://github.com/zephyrproject-rtos/zephyr/blob/master/include/net/net_pkt.h#L1145
In net_pkt, void * are only used for user_data and token storage.

That being said, net_buf APIs are fairly varied throughout.  Sometimes using void * for data and sometimes using u8_t *.  For instance:
`static inline u8_t *net_buf_simple_tail(struct net_buf_simple *buf)`
returns the tail pointer for the buffer as a u8_t *.
But, `void *net_buf_simple_add(struct net_buf_simple *buf, size_t len)` returns "The original tail of the buffer" as a void *.

@jukkar have a preference?

Reply:
I am ok with both.

Reply:
Changing to `void *`